ui/src/dashboards/components/VEO.tsx

```diff
@@ -1,87 +1,153 @@
+  onUpdateView: typeof viewActions.updateView
+  notify: typeof notify
+  setActiveTimeMachine: typeof setActiveTimeMachine
 }
 
-interface OwnProps {
-  onHide: () => void
-  onSave: (v: NewView | View) => Promise<void>
+interface OwnProps extends WithRouterProps {
+  params: {
+    dashboardID: string
+    cellID?: string
+  }
 }
 
 type Props = OwnProps & StateProps & DispatchProps
 
 class VEO extends PureComponent<Props, {}> {
   public render() {
-    const {draftView, onSetName, onHide} = this.props
+    const {draftView, onSetName} = this.props
 
     return (
-      <div className="veo">
-        <VEOHeader
-          key={draftView.name}
-          name={draftView.name}
-          onSetName={onSetName}
-          onCancel={onHide}
-          onSave={this.handleSave}
-        />
-        <div className="veo-contents">
-          <TimeMachine />
+      <Overlay visible={true} className="veo-overlay">
+        <div className="veo">
+          <VEOHeader
+            key={draftView.name}
+            name={draftView.name}
+            onSetName={onSetName}
+            onCancel={this.close}
+            onSave={this.handleSave}
+          />
+          <div className="veo-contents">
+            <TimeMachine />
+          </div>
         </div>
-      </div>
+      </Overlay>
     )
   }
 
-  private handleSave = (): void => {
-    const {draftView, draftQueries, onSave} = this.props
+  public async componentDidMount() {
+    const {existingView} = this.props
+    const view = existingView || createView<XYView>(ViewType.XY)
+
+    this.props.setActiveTimeMachine(VEO_TIME_MACHINE_ID, {view})
+  }
 
-    // Ensure that the latest queries are saved with the view, even if they
-    // haven't been submitted yet
-    const view = {
+  private handleSave = async (): Promise<void> => {
+    const {
+      draftView,
+      draftQueries,
+      dashboard,
+      onCreateCellWithView,
+      onUpdateView,
+      notify,
+    } = this.props
+
+    const view: QueryView & {id?: string} = {
       ...draftView,
       properties: {
         ...draftView.properties,
         queries: draftQueries,
       },
     }
 
-    onSave(view)
+    try {
+      if (view.id) {
+        await onUpdateView(dashboard.id, view)
+      } else {
+        await onCreateCellWithView(dashboard, view)
+      }
+      this.close()
+    } catch (error) {
+      console.error(error)
+      notify(cellAddFailed())
+    }
+  }
+
+  private close = (): void => {
+    this.props.router.goBack()
   }
 }
 
-const mstp = (state: AppState): StateProps => {
+const mstp = (state: AppState, {params}): StateProps => {
+  const {cellID, dashboardID} = params
+  const {
+    views: {views},
+    dashboards,
+  } = state
   const {view, draftQueries} = getActiveTimeMachine(state)
+  const dashboard = dashboards.find(d => d.id === dashboardID)
+  let existingView = null
+  if (cellID) {
+    existingView = (views[cellID] || {view: null}).view
```

Comment:
can this be views[cellID].view || null instead? hard to read

Reply:
no because if `views[cellID]` returns null then `views[cellID].view` will throw an error

Reply:
hrmm _.get?